**UNIVERSIDADE LUSÓFONA**

# Programação Web -  Ficha 2: Python

* Estes exercícios estão redigidos em forma de tutorial para aprender e exercitar conceitos chave do Python. 
* Consulte os slides se tiver dúvidas. Estes contêm toda a matéria necessária para realizar estes exercícios.
* Consulte os [video-tutoriais](https://educast.fccn.pt/vod/channels/1fzfn2fjbd?locale=en) curtos que explicam conceitos específicos do Python.


# 1 Listas de números
Considere a seguinte lista (execute a linha de código seguinte para que a lista fique definida no Jupyter)

In [19]:
lista = [1, 4, 6, 22, 4, 7, 13, 12, 2, 3, 5]

In [20]:
print(lista[0])

1


Lembre-se que existe:
* uma indexação que vai da esquerda para a direita e começa no 0, 
* uma indexação que vai da direita para a esquerda e começa no -1  



1.1 imprima o numero que está na posição de indice 2

In [21]:
print(lista[2])

6


1.2 imprima o numero que está na penúltima posição da lista.

In [22]:
print(lista[-2])

3


1.3 imprima todos os números desta lista separados por espaços (use o atributo `end` da função `print`). Lembre-se que com um ciclo `for` pode iterar diretamente pelos elementos da lista.

In [23]:
for pos in range(len(lista)):
    print(lista[pos], end=" ")

1 4 6 22 4 7 13 12 2 3 5 

Criemos uma lista dos números de 20 a 30. Imprimamos a posição e valor de cada elemento da lista

In [24]:
lista_numeros = []  
for numero in range(20, 30+1):   
    lista_numeros.append(numero)
    
for i, n in enumerate(lista_numeros):   # enumerate retorna i índice e o valor 😀
    print(f"posição {i:02}: {n}")   # 02 de {i:02} indica que o numero deve ocupar 2 casas, com 0 se não houver digito

posição 00: 20
posição 01: 21
posição 02: 22
posição 03: 23
posição 04: 24
posição 05: 25
posição 06: 26
posição 07: 27
posição 08: 28
posição 09: 29
posição 10: 30


1.4. Recorrendo a um ciclo `for`, faça um programa que 
1. some todos os números pares da variável `lista` 
1. crie a lista pares, dos elementos pares da lista
1. imprima o resultado e a lista pares.

Resultado esperado:
```
A soma dos numeros pares da lista é 50
Lista dos números pares: [4, 6, 22, 4, 12, 2]
```

In [25]:
lista = [1, 4, 6, 22, 4, 7, 13, 12, 2, 3, 5]
total = 0
lista_pares = []
for pos in lista:
    if pos % 2 == 0:
        total += pos
        lista_pares.append(pos)
print(f"A soma dos numeros pares da lista é {total}")
print(f"Lista dos números pares: {lista_pares}")


A soma dos numeros pares da lista é 50
Lista dos números pares: [4, 6, 22, 4, 12, 2]


1.5 Faça um programa que imprime o índice da posição e o valor de cada elemento de `lista`. Utilize a função `enumerate`.

resultado esperado:
```
Na posição de indice 0 está o 1
Na posição de indice 1 está o 4
...
```

In [26]:
for indice, valor in enumerate(lista):
    print(f"Na posição de indice {indice} está o {valor}")

Na posição de indice 0 está o 1
Na posição de indice 1 está o 4
Na posição de indice 2 está o 6
Na posição de indice 3 está o 22
Na posição de indice 4 está o 4
Na posição de indice 5 está o 7
Na posição de indice 6 está o 13
Na posição de indice 7 está o 12
Na posição de indice 8 está o 2
Na posição de indice 9 está o 3
Na posição de indice 10 está o 5


Crie uma lista com os números de 1 a 10 usando uma compreensão

In [27]:
lista = [i for i in range(1,10+1)]  # isto é uma compreensão, uma forma "pythoniana" de criar uma lista 😉

1.6 Crie uma lista com os números de 100 a 200 usando uma compreensão

In [28]:
numeros_100_200 = [i for i in range(100, 201)]
print(numeros_100_200)

[100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200]


1.7 Cria uma lista com os números pares de 100 a 140, inclusivé,  usando uma compreensão

In [29]:
pares_100_140 = [i for i in range(100, 141) if i % 2 == 0]
print(pares_100_140)

[100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140]


1.8 Ordene a `lista` crescentemente com o método `sort`, e imprima-a. Deve fazer em duas instruções separadas, primeiro ordene, e depois imprimir. 

In [30]:
lista = [1, 4, 6, 22, 4, 7, 13, 12, 2, 3, 5]
lista.sort()
print(lista)

[1, 2, 3, 4, 4, 5, 6, 7, 12, 13, 22]


1.9  
* Faça uma função que pede ao utilizador um número inteiro positivo. (Utilize o `try except` para garantir que não há erros por não inserir um inteiro). 
* Utilize a função para preencher uma lista apenas com os números que sejam pares.  
* Ordene a lista de números, indique o maior, o menor, e o que ocorre mais vezes.

In [39]:
import statistics
lista = list() 
while(True):
    try:
        num = int(input("Insira um número inteiro positivo "))
        if num < 0: 
            break
        if num % 2 == 0:
            lista.append(num)
    except:
        break
lista.sort()
print(max(lista))
print(min(lista))
print(statistics.mode(lista))

454
4
6


# 2 Listas de strings

Considere a seguinte lista: 

`nomes = ["Maria", "Manuel", "Luis", "Augusta", "Zé", "Tolentino", "Ana"]`



2.1 Recorrendo a um ciclo e a formated strings, itere pelos vários nomes e imprima para cada nome uma frase do tipo:

`Bom dia, Maria!`

In [40]:
nomes = ["Maria", "Manuel", "Luis", "Augusta", "Zé", "Tolentino", "Ana"]

for nome in nomes:
    print(f"Bom dia, {nome}!")

Bom dia, Maria!
Bom dia, Manuel!
Bom dia, Luis!
Bom dia, Augusta!
Bom dia, Zé!
Bom dia, Tolentino!
Bom dia, Ana!


2.2 Crie uma lista ordenada alfabeticamente com o método sort, e imprima-a.

Resultado esperado: `['Ana', 'Augusta', 'Luis', 'Manuel', 'Maria', 'Tolentino', 'Zé']`

In [41]:
nomes = ["Maria", "Manuel", "Luis", "Augusta", "Zé", "Tolentino", "Ana"]

# ordenação da lista alfabeticamente

nomes.sort()

print(nomes)

# Resultado esperado: ['Ana', 'Augusta', 'Luis', 'Manuel', 'Maria', 'Tolentino', 'Zé']

['Ana', 'Augusta', 'Luis', 'Manuel', 'Maria', 'Tolentino', 'Zé']


2.3 Crie uma lista ordenada pelo número de carateres, de forma decrescente, e imprima-a. Use o atributo key e um lambda, tal como foi feito na aula.

Resultado esperado: `['Tolentino', 'Augusta', 'Manuel', 'Maria', 'Luis', 'Ana', 'Zé']`

In [46]:
nomes = ["Maria", "Manuel", "Luis", "Augusta", "Zé", "Tolentino", "Ana"]
nomes_ordenados = sorted(nomes, key=lambda nome: len(nome), reverse=True)
print(nomes_ordenados)

['Tolentino', 'Augusta', 'Manuel', 'Maria', 'Luis', 'Ana', 'Zé']


2.4 Crie, com um ciclo for, uma nova lista, `lista_letras`, com a primeira letra de cada nome. Imprima-a (pode fazer diretamente `print(lista_letras)`)

Resultado esperado: `['M', 'M', 'L', 'A', 'Z', 'T', 'A']`

In [47]:
nomes = ["Maria", "Manuel", "Luis", "Augusta", "Zé", "Tolentino", "Ana"]
lista_letras = []
for nome in nomes:
    lista_letras.append(nome[0])

print(lista_letras)

['M', 'M', 'L', 'A', 'Z', 'T', 'A']


2.5 Usando a lista `nomes`, crie com uma compreensão uma nova lista que guarda a primeira letra de cada nome.

In [48]:
nomes = ["Maria", "Manuel", "Luis", "Augusta", "Zé", "Tolentino", "Ana"]

lista_letras = [nome[0] for nome in nomes]
print(lista_letras)

['M', 'M', 'L', 'A', 'Z', 'T', 'A']


2.6 Crie a lista `ladies` que itera pela lista `nomes` e guarda apenas os nomes terminados em "a".


Resultado esperado: `['Maria', 'Augusta', 'Ana']`

In [50]:
nomes = ["Maria", "Manuel", "Luis", "Augusta", "Zé", "Tolentino", "Ana"]

ladies = [nome for nome in nomes if nome.endswith("a")]
print(ladies)

['Maria', 'Augusta', 'Ana']


2.7 Crie a lista `ladies` que guarda os nomes terminados em "a", usando uma compreensão

In [51]:
nomes = ["Maria", "Manuel", "Luis", "Augusta", "Zé", "Tolentino", "Ana"]

ladies = [nome for nome in nomes if nome[-1] == "a"]
print(ladies)

['Maria', 'Augusta', 'Ana']


# 3 Processamento de listas com strings

### Relembrando...

No processamento de strings, dois métodos são muito úteis, split e find. Veja e experimente os exemplos em baixo:

In [32]:
"Pedro Tamen, 1/1/1934".split(', ')  # retorna uma lista de elementos separados por ', '

['Pedro Tamen', '1/1/1934']

In [33]:
"Pedro Tamen, 1/1/1934".split(',')[0]  # podemos aceder ao primeiro elemento da lista, o nome completo

'Pedro Tamen'

In [34]:
# extraia o apelido da variável em baixo

nome = "Pedro Tamen, 1/1/1934"
nome.split()[1][:-1]

'Tamen'

As compreensões também são formas simples de transformar uma lista. Por exemplo, criemos a lista com a primeira letra de cada pessoa: 

In [35]:
dados = [
    "Pedro Tamen, 1/1/1934", 
    "Almeida Garrett, 4/2/1799",
    "Camilo Pessanha, 7/9/1867",
    "Almada Negreiros, 7/5/1893",
    "Ibn Bassam, 21/11/1058",
    "Antonio Aleixo, 18/2/1899",
    "Agustina Bessa-Luís, 15/10/1922",
    "Ricardo Reis, 19/9/1887",
    "Mario Sá-Carneiro, 19/5/1890",
    "Mario Cesariny, 9/9/1923",
    "Luis Camões, 10/6/1524",
    "Miguel Torga, 12/8/1907",
    "Lidia Jorge, 18/10/1946",
    "Natália Correia, 12/9/1923",
    "Tolentino Mendonça, 15/12/1965",
    "Alberto Caeiro, 16/5/1889",
    "Jose Saramago, 16/11/1922",
    "Cesario Verde, 25/2/1855",
    "Luiz Pacheco, 7/5/1925",
    "Manuel Alegre, 12/5/1936",
    "Fernando Pessoa, 13/6/1888",
    "Herberto Helder, 23/11/1930",
    "Florbela Espanca, 8/12/1894"
]

letras = [info[0] for info in dados]
print(letras)

['P', 'A', 'C', 'A', 'I', 'A', 'A', 'R', 'M', 'M', 'L', 'M', 'L', 'N', 'T', 'A', 'J', 'C', 'L', 'M', 'F', 'H', 'F']


### Exercicios 
Considere a seguinte lista de escritores portugueses e suas datas de nascimento:

In [36]:
dados = [
    "Pedro Tamen, 1/1/1934", 
    "Almeida Garrett, 4/2/1799",
    "Camilo Pessanha, 7/9/1867",
    "Almada Negreiros, 7/5/1893",
    "Ibn Bassam, 21/11/1058",
    "Antonio Aleixo, 18/2/1899",
    "Agustina Bessa-Luís, 15/10/1922",
    "Ricardo Reis, 19/9/1887",
    "Mario Sá-Carneiro, 19/5/1890",
    "Mario Cesariny, 9/9/1923",
    "Luis Camões, 10/6/1524",
    "Miguel Torga, 12/8/1907",
    "Lidia Jorge, 18/10/1946",
    "Natália Correia, 12/9/1923",
    "Tolentino Mendonça, 15/12/1965",
    "Alberto Caeiro, 16/5/1889",
    "Jose Saramago, 16/11/1922",
    "Cesario Verde, 25/2/1855",
    "Luiz Pacheco, 7/5/1925",
    "Manuel Alegre, 12/5/1936",
    "Fernando Pessoa, 13/6/1888",
    "Herberto Helder, 23/11/1930",
    "Florbela Espanca, 8/12/1894"
]

3.1 Crie, com um ciclo for, uma lista de strings, onde cada string tem o nome e apelido de cada pessoa. Utilize por exemplo o método split.

   * Exemplo de excerto da lista: `nomes = ["Pedro Tamen", "Almeida Garrett", ...]` 

In [59]:
nomes = []
for linha in dados:
    nome = linha.split(", ")[0]
    nomes.append(nome)

print(nomes)

['Pedro Tamen', 'Almeida Garrett', 'Camilo Pessanha', 'Almada Negreiros', 'Ibn Bassam', 'Antonio Aleixo', 'Agustina Bessa-Luís', 'Ricardo Reis', 'Mario Sá-Carneiro', 'Mario Cesariny', 'Luis Camões', 'Miguel Torga', 'Lidia Jorge', 'Natália Correia', 'Tolentino Mendonça', 'Alberto Caeiro', 'Jose Saramago', 'Cesario Verde', 'Luiz Pacheco', 'Manuel Alegre', 'Fernando Pessoa', 'Herberto Helder', 'Florbela Espanca']


3.2 Crie uma lista de strings, onde cada string tem a primeira letra (abreviatura) do nome e o apelido de cada pessoa. Tende depois utilizar uma compreensão.

   * Exemplo de excerto da lista:  `nomes = ["P.Tamen", "A.Garrett", ...]` 

In [65]:
nomes = [linha.split(", ")[0].split()[0][0] + "." + linha.split(", ")[0].split()[-1] for linha in dados]
print(nomes)

['P.Tamen', 'A.Garrett', 'C.Pessanha', 'A.Negreiros', 'I.Bassam', 'A.Aleixo', 'A.Bessa-Luís', 'R.Reis', 'M.Sá-Carneiro', 'M.Cesariny', 'L.Camões', 'M.Torga', 'L.Jorge', 'N.Correia', 'T.Mendonça', 'A.Caeiro', 'J.Saramago', 'C.Verde', 'L.Pacheco', 'M.Alegre', 'F.Pessoa', 'H.Helder', 'F.Espanca']


3.3 Ordene a lista pelo apelido

Resultado esperado: `['M.Alegre', 'A.Aleixo', 'I.Bassam', 'A.Bessa-Luís', ...`

In [63]:
nomes = [linha.split(", ")[0].split()[0][0] + "." + linha.split(", ")[0].split()[-1] for linha in dados]
nomes_ordenados = sorted(nomes, key=lambda nome: nome.split(".")[1])
print(nomes_ordenados)

['M.Alegre', 'A.Aleixo', 'I.Bassam', 'A.Bessa-Luís', 'A.Caeiro', 'L.Camões', 'M.Cesariny', 'N.Correia', 'F.Espanca', 'A.Garrett', 'H.Helder', 'L.Jorge', 'T.Mendonça', 'A.Negreiros', 'L.Pacheco', 'C.Pessanha', 'F.Pessoa', 'R.Reis', 'J.Saramago', 'M.Sá-Carneiro', 'P.Tamen', 'M.Torga', 'C.Verde']
